cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/AbstractConnectOperator.java

```diff
@@ -565,6 +565,7 @@ public static void updateStatus(Throwable error, KafkaConnector kafkaConnector2,
         }
         StatusUtils.setStatusConditionAndObservedGeneration(connector, status, error != null ? Future.failedFuture(error) : Future.succeededFuture());
         status.setConnectorStatus(statusResult);
+        status.setTasksMax(connector.getSpec().getTasksMax());
```

Comment:
This isn't correct. A `Connector.taskConfigs(int maxTasks)` is called by Connect to get the tasks configs, but the connector is not obliged to return `maxTasks` configs. Some connectors (such as the Debezium [MySQL connector}(https://debezium.io/documentation/reference/1.1/connectors/mysql.html)) return fewer task configs. So we would at least need to get the actual number of tasks that the connector is using, via the REST API, rather than copying the value from the spec.

Something else to be aware of: for a source connector it's possible that returning >1 task config when there's no way of partitioning the source data will result in duplicated records in the topic, because two or more tasks consume the same events from the source system. In theory people should be writing their connectors to return only 1 task config in that case, but that doesn't mean people actually do that in practice. Turning an autoscaler loose on such connectors would result in dupes. So it might be worthwhile have some way, in the CR, of opting in to having these published on the `status` (or maybe autoscalers already work by selecting their victims via labels, and it would be up to the user to not configure those labels for `KafkaConnector` resources which didn't properly support multiple tasks.

Reply:
I actually think using the maximum number is better. Otherwise you can in theory run in _loop_ trying to scale more than what is the number of tasks the connector actually runs. So I do not think this is an issue. If you would scale an deployment, it would actually make sure the pods are running, but will also have no control over whether they do something or whether they do duplicate work. So I do not think it is radically different.

In general, all what this PR allows is that you can do something like `kubectl scale <kind> <name> --replicas N`. So on its own this is not different from the user taking the connector and scaling it to N tasks by editing the resource. It does not so any autoscaling, it is just an enabler.

Achieving some form of autoscaling is more complicated and needs some more steps. For the resources which have the LabalSelector (Connect, Mirror Maker, Bridge), the user can create a `HorizontalPodAutoscaler` resource and have it to autoscale the resources based on some metric. So the users needs to do some additional deliberate action where (s)he hopefully considers the consequences. Out of the box, the scaling can be done for example based on CPU utilization - but we know that with Kafka this is not sufficient due to consumer groups etc. So user would really need some additional tooling to deal with it.

With things such as KafkaConnector, the HPA as far as I know cannot be used because it does not have the label selector (and there is really nothing to select since the resource does not live as pod or any other Kubernetes resource). TBH, I'm not sure how much use would the scale subresource get here unless we (or some user) write some specific tooling around it. But it seems like a small effort and something what would not hurt and might enable interesting patterns.

A chapter on its own would be probably something like scaling of Kafka topics which is I guess even more tricky (since you cannot scale down).

Reply:
> Otherwise you can in theory run in loop trying to scale more than what is the number of tasks the connector actually runs.

I'm not sure I understand this loop. But I can see a looping problem when the `task.max` is used as the scale in the `status`:

1. Some actor decides (based on $metrics) that more scale is needed.
2. They increase the scale in the `spec`
3. Because the the connector has some hard coded limit about the number of tasks creates it ignores the `task.max`
4. But the operator updates the `status` to match the `spec`, so it looks like that many tasks have been created.
5. The $metrics in point 1 haven't changed, so the scale is increased again.

Reply:
Well, the loop I would be afraid of is when the scaler decides that the scale should be N and is triggering it all the time while the custom resource is ignoring it because the connector decided it will use only 1 task. The loop you are describing will normally stop at some upper boundary set in the autoscaler.

I'm also not completely sure what your concern is. I can see how the user can screw it up. So can he screw up scaling for Kafka Consumer and many other workloads running in Kubernetes Deployment. Yet I do not think it is a reson to remove the scale resource from Deployments. In this case it is actually a bit esier since the tasks seem to be fairly virtual, so the scaler would stop at some upper limit and not even waste so much resources as with Deployments and Pods for example.

Anyway, one of the issues is that the maxTasks are not mandatory and do not have any default in Strimzi. I'm not yet sure how to deal with it - might decide to just remove the scale subresource from the connector to make my life easier.

Reply:
> the loop I would be afraid of is when the scaler decides that the scale should be N and is triggering it all the time while the custom resource is ignoring it because the connector decided it will use only 1 task. 

I would argue that a scaler should observe that the `status.observedGeneration` matches the `metadata.generation` after the scaling, but the `status.${scale}` does not reflect the `spec.${scale}` and infer that, for whatever reason, the resource can't be scaled that high (or at least, not right now) and thus not try to increase the scale again. I say "I would argue", because until there are actually scalers for CRs like this it's a bit of a moot point. And perhaps that's an argument for not making the `KafkaConnector` scalable at this point. Or perhaps we should open a conversation with the KEDA folks about this. Because it would be annoying if we made what turned out to be the wrong decision here.

> The loop you are describing will normally stop at some upper boundary set in the autoscaler.

Does the autoscale have such a limit and is it mandatory for it to be set?

Reply:
I'm afraid I still don't understand the concerns you have. Yes, some connectors cannot scale. This is not unique for connectors, there are many applications which do not scale and yet they have the scale subresource in their deployments and just don't use it. This doesn't add any new features - it just makes it easier to change the `.spec.tasksMax` field using the scale subresource. But even today a user can go and set `tasksMax` to 10 for a connector which can do only one task. So for me this is something what can help some users and will not hurt others. Once again, this does not do any autoscaling unless user plugs it into some different tooling - so it will not scale to milion tasks for anyone out of the box. So what is the issue you expect?
 
> I would argue that a scaler should observe that the status.observedGeneration matches the metadata.generation after the scaling, but the status.${scale} does not reflect the spec.${scale} and infer that, for whatever reason, the resource can't be scaled that high (or at least, not right now) and thus not try to increase the scale again.

Not sure what `observedGeneration` has to do with it. It changes one way or another. I also after my investigation don't think you can get any better number of tasks. The connector status is using very ambigous and doesn't for example allow you to determine if `UNASSIGNED` tasks are scaled down or if they are just temporarily unassigned because of node restarts. So for example counting all tasks in status or counting running tasks does not give any reasonable value either.

>  I say "I would argue", because until there are actually scalers for CRs like this it's a bit of a moot point. And perhaps that's an argument for not making the KafkaConnector scalable at this point.

I think this is very unfriendly to the ecosystem. If you wait for the tooling to exist, you basically wait for someone to implement it in a much more complicated way (by hardcoding the mechanism for editing KadkaConnector CR and changing tasksMax number). And then you would implement the scale subresource and tell them _hey, you can do this in a much easier way_.

> Does the autoscale have such a limit and is it mandatory for it to be set?

I do not think this question has a single answer. AFAIK HorizontalPodAutoscaler has `maxReplicas` as required value. But there are many ways how to implement autoscaling and I do not know if it applies to all of them. And as I said, HorizontalPodAutoscaler does not apply to KafkaConnector but only to pod based resources.

Reply:
> I'm afraid I still don't understand the concerns you have.

I think you're misunderstanding my point. I'm not against adding scale subresource to `KafkaConnector` at all. I'm just trying to understand the consequences of having a scale in the `status` that does not reflect the true scale at which the connector be being run. In general having system A lie to system B about something which system B will use to automated system A seems like it could go wrong, so I want to be pretty sure that there are no negative consequences, as you assert.

> there are many applications which do not scale and yet they have the scale subresource in their deployments and just don't use it.

Well, if there's an established pattern of having the scale in the status not actually reflect the true scale, because it's being ignored then I guess that's fine. But it would be good to know which operators are already doing this.

> Not sure what observedGeneration has to do with it.

By comparing the `observedGeneration` with the `metadata.generation` the scaler can infer whether the change in scale that was requests has been actioned by the operator. If the generations match and the status scale is not the spec scale then the scaler can infer that the scaling didn't happen for some reason.

> If you wait for the tooling to exist, you basically wait for someone to implement it in a much more complicated way

I'm not saying we have to wait, as such. But since this is the interface that any such tooling will use I do think it's important that we implement something which will actually work when that tooling exists.

> there are many ways how to implement autoscaling and I do not know if it applies to all of them

So that means that "my" loop is indeed potentially an issue.